File: ml-parcial-IIND3118/ml-parcial-IIND3118/temas/B09_mds/python/B09_mds.py
"""
B9. MDS (Multidimensional Scaling)
Tema B09_mds — Introducción al Machine Learning (IIND3118)

CÓMO USARLO
  1. Cambia el bloque marcado con ✏️ abajo por tu dataset.
  2. Corre el archivo:  python B09_mds.py
  3. Las gráficas se guardan en la carpeta figuras/ de este tema.

La explicación del método está en el README.md de esta misma carpeta.
"""

import os, sys, warnings
warnings.filterwarnings("ignore")

import numpy as np
import pandas as pd
import matplotlib
matplotlib.use("Agg")          # no abre ventanas: funciona en cualquier entorno
import matplotlib.pyplot as plt

from sklearn.model_selection import train_test_split, KFold, cross_val_score, GridSearchCV
from sklearn.preprocessing import scale
from sklearn.linear_model import (LinearRegression, Ridge, Lasso, ElasticNet,
                                  RidgeCV, LassoCV, ElasticNetCV, LogisticRegression)
from sklearn.neighbors import KNeighborsRegressor, KNeighborsClassifier
from sklearn.metrics import (mean_squared_error, mean_absolute_error, r2_score,
                             accuracy_score, confusion_matrix, classification_report,
                             roc_curve, roc_auc_score)
from sklearn.decomposition import PCA, KernelPCA
from sklearn.cluster import KMeans, DBSCAN
from sklearn.ensemble import IsolationForest
from sklearn.neighbors import LocalOutlierFactor
from sklearn import manifold

# ── Guardado automático de figuras ──────────────────────────────────────
AQUI = os.path.dirname(os.path.abspath(__file__))
FIGS = os.path.join(AQUI, "..", "figuras")
os.makedirs(FIGS, exist_ok=True)
_n = [0]
def _save():
    _n[0] += 1
    ruta = os.path.join(FIGS, f"{_n[0]:02d}_figura.png")
    plt.savefig(ruta, dpi=110, bbox_inches="tight"); plt.close()
    print(f"   [figura guardada: figuras/{os.path.basename(ruta)}]")
plt.show = _save                          # cada plt.show() guarda en vez de abrir

RNG = np.random.default_rng(0)

# ╔═════════════════════════════════════════════════════════════════════════╗
# ║  ✏️  CAMBIA SOLO ESTE BLOQUE POR TU DATASET                             ║
# ╚═════════════════════════════════════════════════════════════════════════╝
DATOS  = os.path.join(AQUI, "..", "..", "..", "data", "ejemplo_regresion.csv")
TARGET = "target"          # nombre EXACTO de la columna respuesta. None si no hay.
# ╚═════════════════════════════════════════════════════════════════════════╝

df = pd.read_csv(DATOS).dropna()

if TARGET is not None:
    y_serie = df[TARGET]; X_df = df.drop(columns=[TARGET])
else:
    y_serie = None; X_df = df.copy()

X_df = pd.get_dummies(X_df, drop_first=True).astype(float)
nombres_X = list(X_df.columns)
X  = X_df.values
Xs = scale(X)

if y_serie is not None:
    y = pd.factorize(y_serie)[0] if y_serie.dtype == object else y_serie.values
    ES_CLASIFICACION = (len(np.unique(y)) <= 10) and np.all(y == np.round(y))
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=.30, random_state=0)
    Xs_train, Xs_test, _, _ = train_test_split(Xs, y, test_size=.30, random_state=0)
else:
    y = None; ES_CLASIFICACION = False

def ModeloLineal(**kw):
    return (LogisticRegression(max_iter=5000, **kw) if ES_CLASIFICACION
            else LinearRegression(**kw))

col = y if y is not None else "steelblue"
Z   = PCA().fit_transform(Xs)        # proyección PCA, se usa para graficar
Z2  = Z[:, :2]

print("="*74)
print("B9. MDS (Multidimensional Scaling)")
print("="*74)
print(f"datos: {X.shape[0]} obs. x {X.shape[1]} predictores")
if y is not None:
    print(f"y ('{TARGET}'): " + ("CLASIFICACIÓN" if ES_CLASIFICACION else "REGRESIÓN"))
print()

# ════════════════════════════════════════════════════════════════════════
#  CÓDIGO DEL MÉTODO
# ════════════════════════════════════════════════════════════════════════

# ═══ B9. MDS ═══
from sklearn.manifold import MDS
from scipy.spatial.distance import pdist, squareform

def cmdscale(D, k=2):
    """MDS clásico (= cmdscale de R). Devuelve (coordenadas, autovalores)."""
    n = D.shape[0]
    C = np.eye(n) - np.ones((n,n))/n          # matriz de centrado
    B = -0.5 * C @ (D**2) @ C                 # doble centrado
    val, vec = np.linalg.eigh(B)
    o = np.argsort(val)[::-1]; val, vec = val[o], vec[:,o]
    pos = val[:k] > 0
    coords = np.zeros((n,k)); coords[:,pos] = vec[:,:k][:,pos]*np.sqrt(val[:k][pos])
    return coords, val

D = squareform(pdist(Xs, metric="euclidean"))   # ⬅ cambia la métrica si quieres
pts, val = cmdscale(D, k=2)

print(f"autovalores (primeros 5): {np.round(val[:5],3)}")
print(f"negativos relevantes    : {int((val<-1e-8).sum())}  (0 => D SÍ es euclidiana)")
print(f"varianza explicada 2D   : {val[:2].sum()/val[val>0].sum():.2%}")

# Verificación: MDS clásico == PCA
print(f"\nMDS clásico vs PCA, diferencia máxima: "
      f"{np.abs(np.abs(pts)-np.abs(Z[:,:2])).max():.2e}  <- deben ser iguales")

# MDS no métrico (solo usa el ORDEN de las disimilitudes)
mds_nm = MDS(n_components=2, metric=False, dissimilarity="precomputed",
             random_state=0, n_init=4).fit_transform(D)

def stress(D_, p_):
    o, e = squareform(D_, checks=False), pdist(p_)
    s = np.sum(o*e)/np.sum(e**2)
    return np.sqrt(np.sum((o-s*e)**2)/np.sum(o**2))
print(f"STRESS-1 (no métrico): {stress(D, mds_nm):.4f}  (<0.10 es bueno)")

fig, ax = plt.subplots(1, 3, figsize=(16, 4.2))
ax[0].scatter(pts[:,0], pts[:,1], c=col, cmap="viridis", alpha=.7); ax[0].set_title("MDS clásico (= PCA)")
ax[1].scatter(mds_nm[:,0], mds_nm[:,1], c=col, cmap="viridis", alpha=.7); ax[1].set_title("MDS no métrico")
o, e = squareform(D, checks=False), pdist(pts)
ax[2].scatter(o, e, alpha=.15, s=6); lim=[0,max(o.max(),e.max())]
ax[2].plot(lim, lim, "r--"); ax[2].set_xlabel("disimilitud original")
ax[2].set_ylabel("distancia en el mapa")
ax[2].set_title(f"Diagrama de Shepard (r={np.corrcoef(o,e)[0,1]:.3f})")
plt.tight_layout(); plt.show()
print("Shepard: entre más pegados a la diagonal, mejor preserva las distancias.")

print("\nListo. Revisa la carpeta figuras/")

File: ml-parcial-IIND3118/ml-parcial-IIND3118/temas/B09_mds/R/B09_mds.R
# ############################################################################
# B9. MDS
# Tema B09_mds — Introducción al Machine Learning (IIND3118)
#
# CÓMO USARLO
#   1. Cambia el bloque marcado con ✏️ por tu dataset.
#   2. Corre:  Rscript B09_mds.R     (o ábrelo en RStudio)
#   3. Las gráficas se guardan en la carpeta figuras/ de este tema.
#
# La explicación del método está en el README.md de esta misma carpeta.
# ############################################################################

for (p in c("MASS","cluster","dbscan","glmnet","leaps","caret","class","fpc",
            "clusterCrit","isotree","kernlab","Rtsne","psych","car","vegan")) {
  if (!requireNamespace(p, quietly=TRUE))
    try(install.packages(p, repos="https://cloud.r-project.org"), silent=TRUE)
}
suppressMessages({ library(MASS); library(cluster) })

set.seed(0)

# Las gráficas se guardan en figuras/ en vez de abrirse en pantalla
dir.create("../figuras", showWarnings=FALSE)
.n <- 0
guardar <- function() {
  .n <<- .n + 1
  dev.copy(png, sprintf("../figuras/%02d_figura.png", .n), width=900, height=650)
  invisible(dev.off())
}

# ╔═══════════════════════════════════════════════════════════════════════╗
# ║  ✏️  CAMBIA SOLO ESTE BLOQUE POR TU DATASET                           ║
# ╚═══════════════════════════════════════════════════════════════════════╝
DATOS  <- "../../../data/ejemplo_regresion.csv"
TARGET <- "target"        # nombre EXACTO de la columna respuesta. NULL si no hay.
# ╚═══════════════════════════════════════════════════════════════════════╝

df <- na.omit(read.csv(DATOS))

if (!is.null(TARGET)) {
  y_raw <- df[[TARGET]]
  X_df  <- df[, setdiff(names(df), TARGET), drop=FALSE]
} else { y_raw <- NULL; X_df <- df }

X_df <- as.data.frame(model.matrix(~ . - 1, data=X_df))
nombres_X <- names(X_df)
X  <- as.matrix(X_df)
Xs <- scale(X)

if (!is.null(y_raw)) {
  ES_CLASIFICACION <- is.factor(y_raw) || is.character(y_raw) || length(unique(y_raw)) <= 10
  y <- if (ES_CLASIFICACION) as.factor(y_raw) else as.numeric(y_raw)
  idx <- sample(seq_len(nrow(X)), floor(0.70*nrow(X)))
  Xtrain <- X[idx,]; Xtest <- X[-idx,]; ytrain <- y[idx]; ytest <- y[-idx]
  Xs_train <- Xs[idx,]; Xs_test <- Xs[-idx,]
} else { y <- NULL; ES_CLASIFICACION <- FALSE }

pca  <- prcomp(X, center=TRUE, scale.=TRUE)
Z2   <- pca$x[, 1:2]
col_pt <- if (!is.null(y)) as.integer(as.factor(y)) else "steelblue"

cat(strrep("=",74), "\n")
cat("B9. MDS\n")
cat(strrep("=",74), "\n")
cat(sprintf("datos: %d obs. x %d predictores\n\n", nrow(X), ncol(X)))

# ############################################################################
#  CÓDIGO DEL MÉTODO
# ############################################################################

# ============================================================================
# B9. MDS (MULTIDIMENSIONAL SCALING)
# ----------------------------------------------------------------------------
# QUÉ HACE: tienes SOLO LAS DISTANCIAS entre observaciones, no sus
# coordenadas. MDS RECONSTRUYE UN MAPA de puntos que respete esas distancias.
# El ejemplo clásico: con las distancias por carretera entre ciudades
# europeas, MDS DIBUJA EL MAPA DE EUROPA sin haber visto una latitud.
#
#   Doble centrado: B = -1/2 C D^2 C   con C = I - (1/n)11'
#   Luego descomposición espectral de B.
#
# ⚠️ EQUIVALENCIA CLAVE (la preguntan)
#   MDS CLÁSICO CON DISTANCIA EUCLIDIANA = PCA
#   Mismo procedimiento visto desde dos lados: PCA descompone la covarianza
#   (p x p), MDS la de productos punto (n x n).
#
# ⚠️ AUTOVALORES NEGATIVOS = DIAGNÓSTICO
#   Si aparecen y son grandes, la matriz D NO ES EUCLIDIANA: ninguna
#   configuración de puntos puede reproducir esas distancias exactamente.
#   Pasa con GOWER y con distancias por carretera.
#
# MDS NO MÉTRICO (Kruskal, isoMDS): solo respeta el ORDEN de las
#   disimilitudes. Para datos ordinales (Likert, rankings).
#   STRESS: <5% excelente · <10% bueno · <20% aceptable
#
# PARÁMETROS
#   cmdscale(d, k=2, eig=TRUE)   k = dimensión; eig=TRUE devuelve autovalores
#   isoMDS(d, k=2)               versión no métrica
#
# ❌ O(n^3) · NO PROYECTA DATOS NUEVOS (PCA sí) ·
#    los ejes NO TIENEN INTERPRETACIÓN (se puede rotar/reflejar libremente)
# ============================================================================

D <- dist(Xs)                                  # ⬅ cambia la métrica si quieres
mds <- cmdscale(D, k=2, eig=TRUE)              # ⬅ eig=TRUE para ver autovalores

cat("autovalores (primeros 5):", round(mds$eig[1:5], 3), "\n")
cat(sprintf("negativos relevantes: %d  (0 => D SÍ es euclidiana)\n",
            sum(mds$eig < -1e-8)))
cat(sprintf("varianza explicada 2D: %.2f%%\n",
            100*sum(mds$eig[1:2])/sum(mds$eig[mds$eig>0])))

# Verificación: MDS clásico == PCA
cat(sprintf("MDS vs PCA, diferencia máxima: %.2e  <- deben ser iguales\n",
            max(abs(abs(mds$points) - abs(pca$x[,1:2])))))

# MDS no métrico (solo usa el ORDEN)
mds_nm <- isoMDS(D, k=2, trace=FALSE)
cat(sprintf("STRESS (no métrico): %.2f%%  (<10%% es bueno)\n", mds_nm$stress))

par(mfrow=c(1,3))
plot(mds$points, pch=16, col=if(!is.null(y)) as.integer(as.factor(y)) else "steelblue",
     xlab="Dim 1", ylab="Dim 2", main="MDS clásico (= PCA)")
plot(mds_nm$points, pch=16, col=if(!is.null(y)) as.integer(as.factor(y)) else "steelblue",
     xlab="Dim 1", ylab="Dim 2", main="MDS no métrico")
sh <- Shepard(D, mds$points)
plot(sh, pch=16, cex=.4, col=rgb(0,0,0,.3),
     xlab="disimilitud original", ylab="distancia en el mapa",
     main="Diagrama de Shepard")
abline(0, 1, col="red", lty=2)
par(mfrow=c(1,1))
# Shepard: entre más pegados a la diagonal, mejor preserva las distancias.

cat("\nListo. Revisa la carpeta figuras/\n")
